src/sign.rs

```diff
@@ -25,23 +27,25 @@ pub struct TransactionData<'a> {
 
 impl<'a> TransactionData<'a> {
     /// Sign and return a raw transaction.
-    pub fn sign(&self, key: &SecretKey, chain_id: Option<u64>) -> Result<Bytes, EthsignError> {
+    pub fn sign(&self, key: &PrivateKey, chain_id: Option<u64>) -> Bytes {
         let mut rlp = RlpStream::new();
         self.rlp_append_unsigned(&mut rlp, chain_id);
 
-        let hash = {
-            let mut output = [0u8; 32];
-            let mut hasher = Keccak::v256();
-            hasher.update(&rlp.as_raw());
-            hasher.finalize(&mut output);
-            output
-        };
+        let hash = hash::keccak256(&rlp.as_raw());
         rlp.clear();
 
-        let sig = key.sign(&hash)?;
-        self.rlp_append_signed(&mut rlp, sig, chain_id);
+        // NOTE: secp256k1 messages for singing must be exactly 32 bytes long
+        //   and not be all `0`s. Because the message being signed here is a 32
+        //   byte hash that is computed from non-`0` data (because of RLP
+        //   encoding prefixes) this should never happen, that means this is
+        //   safe to unwrap here.
```

Comment:
A hash function is supposed to give uniformly random outputs, this includes all zeroes. Or is this a specific property of keccak256 that it will never output all zeroes on non zero data?

Reply:
It is an assumption that I made and am verifying :)

Reply:
Updating the comment. From @fedgiac the chance of this happening is almost 0. I think it is OK to unwrap here.

Reply:
The chance of this happening is as high as any other specific hash but if this is nothing we can handle anyway then we might as well unwrap.
How come such a message cannot be signed? Is this a quirk of how secp256k1 works?

Reply:
I mean, we can handle it (and we used to handle signing errors), its just a useless branch that never gets called really which is impossible to write a unit test for.